File: src/Widgets/Components/CharaTrait.cpp
/**
 * \file
 * \author Victor von Rhein <goliath@caern.de>
 *
 * \section License
 *
 * Copyright (C) 2011 by Victor von Rhein
 *
 * This file is part of SoulCreator.
 *
 * SoulCreator is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SoulCreator is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SoulCreator.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <QDebug>

#include "Parser/StringBoolParser.h"
// #include "Exceptions/Exception.h"
#include "Widgets/Dialogs/MessageBox.h"

#include "CharaTrait.h"


CharaTrait::CharaTrait( QWidget* parent, Trait* trait, Trait* traitStorage ) : TraitLine( parent, trait->name(), trait->value() ) {
	// Vorsicht: Nullzeiger ist immer gefährlich!
	ptr_trait = 0;
	ptr_traitStorage = traitStorage;

	character = StorageCharacter::getInstance();

	setTraitPtr( trait );

	// Falls ich mit der Maus den Wert ändere, muß er auch entsprechend verändert werden.
	connect( this, SIGNAL( valueChanged( int ) ), this, SLOT( setTraitValue( int ) ) );
	connect( this, SIGNAL( textChanged( QString ) ), this, SLOT( setCustomText( QString ) ) );
	connect( this, SIGNAL( typeChanged( cv_AbstractTrait::Type ) ), this, SLOT( hideSpecialtyWidget( cv_AbstractTrait::Type ) ) );
	connect( this, SIGNAL( typeChanged( cv_AbstractTrait::Type ) ), this, SLOT( hideDescriptionWidget() ) );
	connect( this, SIGNAL( specialtiesClicked( bool ) ), this, SLOT( emitSpecialtiesClicked( bool ) ) );
	connect( traitPtr(), SIGNAL( detailsChanged(int)), this, SLOT( unclickButton( int ) ) );

	connect( character, SIGNAL( speciesChanged( cv_Species::SpeciesFlag ) ), this, SLOT( hideTraitIfNotAvailable( cv_Species::SpeciesFlag ) ) );
	connect( traitPtr(), SIGNAL( valueChanged( int ) ), this, SLOT( setValue( int ) ) );

	// Die Signale hier zu verbinden funktioniert offensichtlich nicht. Vielleicht weil einige Fertigkeiten dann noch nicht existieren.
	connect( traitPtr(), SIGNAL( availabilityChanged(bool)), this, SLOT( setEnabled(bool)) );

	if ( !traitPtr()->possibleValues().isEmpty() ) {
		setPossibleValues( traitPtr()->possibleValues() );
	}

	hideSpecialtyWidget( trait->type() );
	hideDescriptionWidget();
}


Trait* CharaTrait::traitPtr() const {
	return ptr_trait;
}

void CharaTrait::setTraitPtr( Trait* trait ) {
	if ( ptr_trait != trait ) {
		ptr_trait = trait;
	}
}


// int CharaTrait2::value() const {
// 	return traitPtr()->value();
// }
// void CharaTrait2::setValue( int val ) {
// 	qDebug() << Q_FUNC_INFO << name() << val << value();
// 	if ( value() != val ) {
// 		TraitLine::setValue( val );
// 	}
// }

void CharaTrait::setTraitValue( int val ) {
// 	qDebug() << Q_FUNC_INFO << name() << val << value();
	if ( traitPtr()->value() != val ) {
		traitPtr()->setValue( val );
	}
}


QString CharaTrait::customText() const {
	return traitPtr()->customText();
}
void CharaTrait::setCustomText( QString txt ) {
	if ( traitPtr()->customText() != txt ) {
		traitPtr()->setCustomText( txt );

		TraitLine::setText( txt );

		emit traitChanged( traitPtr() );
	}
}


cv_AbstractTrait::Type CharaTrait::type() const {
	return ptr_trait->type();
}

void CharaTrait::setType( cv_AbstractTrait::Type type ) {
	if ( ptr_trait->type() != type ) {
		ptr_trait->setType( type );

		emit typeChanged( type );
		emit traitChanged( traitPtr() );
	}
}

cv_AbstractTrait::Category CharaTrait::category() const {
	return ptr_trait->category();
}

void CharaTrait::setCategory( cv_AbstractTrait::Category category ) {
	if ( ptr_trait->category() != category ) {
		ptr_trait->setCategory( category );

		emit traitChanged( traitPtr() );
	}
}

cv_Species::Species CharaTrait::species() const {
	return ptr_trait->species();
}

void CharaTrait::setSpecies( cv_Species::Species species ) {
	if ( ptr_trait->species() != species ) {
		ptr_trait->setSpecies( species );
// 		emit speciesChanged(species);

		emit traitChanged( traitPtr() );
	}
}


bool CharaTrait::custom() const {
	return ptr_trait->custom();
}

void CharaTrait::setCustom( bool sw ) {
	if ( ptr_trait->custom() != sw ) {
		ptr_trait->setCustom( sw );

		emit traitChanged( traitPtr() );
	}
}

void CharaTrait::hideSpecialtyWidget( cv_AbstractTrait::Type type ) {
	if ( type == cv_AbstractTrait::Skill ) {
		hideSpecialties( false );
	} else {
		hideSpecialties( true );
	}
}

void CharaTrait::hideDescriptionWidget() {
	if ( custom() ) {
		hideDescription( false );
	} else {
		hideDescription( true );
	}
}

void CharaTrait::emitSpecialtiesClicked( bool sw ) {
	if ( ptr_traitStorage != 0 ) {
		QList< cv_TraitDetail > listStora = ptr_traitStorage->details();
		QList< cv_TraitDetail > listChara = traitPtr()->details();

// 		qDebug() << Q_FUNC_INFO << traitPtr()->name() << ptr_traitStorage->name() << traitPtr()->details().count() << ptr_traitStorage->details().count();

		for ( int i = 0; i < listStora.count(); i++ ) {
			for ( int j = 0; j < listChara.count(); j++ ) {
				if ( listStora.at( i ).name == listChara.at( j ).name ) {
// 					qDebug() << Q_FUNC_INFO << sw << listStora.at( i ).name << listChara.at( j ).name << listChara.at( j ).value;
					cv_TraitDetail traitDetail = listChara.at( j );
					listStora.replace( i, traitDetail );
				}
			}
		}

		emit specialtiesClicked( sw, name(), listStora );
	}
}
void CharaTrait::unclickButton( int val )
{
// 	qDebug() << Q_FUNC_INFO << val;
	if (val < 1){
		setSpecialtyButtonChecked(false);
		emit specialtiesClicked( false, name(), ptr_traitStorage->details());
	}
}



void CharaTrait::hideTraitIfNotAvailable( cv_Species::SpeciesFlag sp ) {
	if ( species().testFlag( sp ) ) {
		setHidden( false );
	} else {
		setValue( 0 );
		setHidden( true );
	}
}

File: src/Widgets/Components/CharaTrait.py
# -*- coding: utf-8 -*-

"""
\file
\author Victor von Rhein <victor@caern.de>

\section License

Copyright (C) Victor von Rhein, 2011, 2012

This file is part of SoulCreator.

SoulCreator is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

SoulCreator is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with SoulCreator.  If not, see <http://www.gnu.org/licenses/>.
"""




from __future__ import division, print_function

from PySide.QtCore import Signal

from src.Config import Config
from src.Widgets.Components.TraitLine import TraitLine
from src.Debug import Debug




class CharaTrait(TraitLine):
	"""
	\brief Mit den gespeicherten Werten vernetzte Darstellung einer einzigen Eigenschaft auf dem Charakterbogen.
 
	Anders als \ref TraitLine, ist dieses Widget direkt mit der korrespondierenden Eigenschaft in der Klasse \ref StorageCharacter verknüpft. Ändert sich der Wert dort, wird automatisch dieses Widget entsprechend verändert. Gleichermaßen wird \ref StorageCharacter verändert, sollte der Benutzer dieses Widget ändern.
 
	\todo Solange kein Text in der Textbox einer Eigenschaft mit Zusatztext steht, sollte der Wert nicht verändert werden können.
	"""


	specialtiesClicked = Signal(bool, object)
	visibilityChanged = Signal(bool)


	def __init__(self, trait, parent=None):
		TraitLine.__init__(self, trait.name, trait.value, parent)

		self.__trait = trait

		# Falls ich mit der Maus den Wert ändere, muß er auch entsprechend verändert werden.
		self.valueChanged.connect(self.setTraitValue)
		self.textChanged.connect(self.setTraitCustomText)
		#connect( this, SIGNAL( typeChanged( cv_AbstractTrait::Type ) ), this, SLOT( hideSpecialtyWidget( cv_AbstractTrait::Type ) ) );
		#connect( this, SIGNAL( typeChanged( cv_AbstractTrait::Type ) ), this, SLOT( hideDescriptionWidget() ) );
		#connect( this, SIGNAL( specialtiesClicked( bool ) ), this, SLOT( emitSpecialtiesClicked( bool ) ) );
		self.buttonToggled.connect(self.emitSpecialtiesClicked)

		#connect( character, SIGNAL( speciesChanged( cv_Species::SpeciesFlag ) ), this, SLOT( hideTraitIfNotAvailable( cv_Species::SpeciesFlag ) ) );
		self.__trait.valueChanged.connect(self.setValue)
		self.__trait.customTextChanged.connect(self.setText)
		self.__trait.specialtiesChanged.connect(self.setSpecialtiesButtonText)

		#// Die Signale hier zu verbinden funktioniert offensichtlich nicht. Vielleicht weil einige Fertigkeiten dann noch nicht existieren.
		#connect( traitPtr(), SIGNAL( availabilityChanged(bool)), this, SLOT( setEnabled(bool)) );
		self.__trait.availableChanged.connect(self.setEnabled)


	def setTraitValue( self, value ):
		"""
		Wenn der Wert dieses Widgets verändert wird, muß auch der dadurch repräsentierte Wert im Speicher verändert werden. Dies geschieht über diesen Slot.
		"""
		
		#Debug.debug("Eigenschaft {} erhält den Wert {}".format(self.__trait["name"], value))
		if self.__trait.value != value:
			self.__trait.value = value
		#Debug.debug("Eigenschaft {} hat den Wert {}".format(self.__trait["name"], self.__trait["value"]))


	#def __getCustomText() const {
		#return traitPtr()->customText();
	#}
	def setTraitCustomText( self, text ):
		"""
		Legt den Zusatztext fest.

		Dabei wird automatisch der Wert im Speicher aktualisiert und natürlich auch die Anzeige des Widget.
		"""

		if self.__trait.customText != text:
			self.__trait.customText = text

			#emit traitChanged( traitPtr() );


	def emitSpecialtiesClicked(self, sw):
		self.specialtiesClicked.emit(sw, self.__trait)


	def hideOrShowTrait_species(self, species):
		"""
		Versteckt oder zeigt diese Eigenschaft, je nach gewählter Spezies.
		"""

		visible = False
		if (not self.__trait.species or self.__trait.species == species):
			visible = True
			#Debug.debug("Verstecke {}, da Alter {} bzw. Ära {}".format(self.name, age, era))

		self.setVisible(visible)
		self.visibilityChanged.emit(visible)


	def hideOrShowTrait(self, age, era):
		"""
		Versteckt oder zeigt diese Eigenschaft.
		"""

		visible = True
		# Es können nur Eigenschaften versteckt werden, die einen age- bzw. era-Eintrag besitzen.
		if (self.__trait.age and self.__trait.age != age) or (self.__trait.era and self.__trait.era != era):
			visible = False
			#Debug.debug("Verstecke {}, da Alter {} bzw. Ära {}".format(self.name, age, era))

		self.setVisible(visible)
		self.visibilityChanged.emit(visible)


	def setSpecialtiesButtonText(self, specialties):
		#count = len(self.__trait.specialties)
		count = len(specialties)
		self.buttonText = count

